include/swift/DependencyScan/DependencyScanningTool.h

```diff
@@ -24,26 +24,34 @@
 namespace swift {
 namespace dependencies {
 class DependencyScanningTool;
+class DependencyScanDiagnosticCollector;
+
+struct ScanQueryInstance {
+  std::unique_ptr<CompilerInstance> ScanInstance;
+  std::unique_ptr<DependencyScanDiagnosticCollector> ScanDiagnostics;
+};
 
 /// Diagnostic consumer that simply collects the diagnostics emitted so-far
-class DependencyScannerDiagnosticCollectingConsumer : public DiagnosticConsumer {
-public:
-  friend DependencyScanningTool;
-  DependencyScannerDiagnosticCollectingConsumer() {}
-  void reset() { Diagnostics.clear(); }
+class DependencyScanDiagnosticCollector : public DiagnosticConsumer {
 private:
   struct ScannerDiagnosticInfo {
     std::string Message;
     llvm::SourceMgr::DiagKind Severity;
   };
-  
+
   void handleDiagnostic(SourceManager &SM,
                         const DiagnosticInfo &Info) override;
   void addDiagnostic(SourceManager &SM, const DiagnosticInfo &Info);
   std::vector<ScannerDiagnosticInfo> Diagnostics;
   // FIXME: For now, we isolate access to shared state of this object
   // but we really should make sure that it doesn't get shared.
   llvm::sys::SmartMutex<true> ScanningDiagnosticConsumerStateLock;
```

Comment:
Should we use a version that doesn't use a lock to handle the new diagnostics that do not race?

Reply:
Added a `LockingDependencyScanDiagnosticCollector` to be used as the global shared one. Good idea, thanks!